### PR TITLE
Create a rudimentary symbol scanning script

### LINKED ISSUE #22330: OS API usage auditing
@nhorman raised an interesting idea that we adopt some sort of policies on what OS APIs we do and don't use. In order to address this we would really need to start with some kind of visibility into what OS APIs we already do depend on. Having an automated API usage audit mechanism here would allow us to put this into CI and have a CI notice trigger automatically if a PR starts depending on an OS API we have never used before.

For symbols (functions/globals), some kind of objdump-based solution may work. It would need to have a way to exempt specific APIs which are approved. But it would also need to have a way to exempt only specific call sites (e.g. for malloc).

Auditing use of OS macros and structure definitions would be harder and would probably require use of compiler-specific functionality, though this is OK if it is something we just use in CI.

Not necessarily a priority but could be quite worthwhile in future to make our use of OS APIs more disciplined.

### PR DESCRIPTION
**Pull request by nhorman**
We would like to be able to log and audit the symbols we use in openssl so that we might catch when a new platform symbols is referecned

Add such a script (just on unix platforms for now) that gathers the used symbols not belonging to libcrypto or libssl, and compare it to a prior known set of used symbols.  Error out if a new symbol is found

Add this script to the static analysis workflow in CI to capture newly introduced platform symbols

Fixes #22330

### CODE REVIEW

File: util/checkplatformsyms.pl [hunk 1]
```diff
@@ -0,0 +1,74 @@
+#! /usr/bin/env perl
+# Copyright 2006-2020 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (slontis) [inline util/checkplatformsyms.pl:2]: Update copyright..

**Comment by t8m** [comment]
I think this is good enough for start. We can improve this later if needed.

**Review by t8m — APPROVED** [review]

**Review by t8m — APPROVED** [review]

File: util/checkplatformsyms.pl [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		$cmd = $cmd . " sed -e\"s/@.*\$//\" | sort | uniq";
+
+		open $expsyms, '<', $expectedsyms or die;
+        {
+            local $/;
+            $exps=<$expsyms>;
+        }
+		close($expsyms);
+
+		open($OBJFH, "$cmd|") or die "Cannot opern process: $!";
```
Comment (hlandau) [inline util/checkplatformsyms.pl:74]: nit: open

**Review by hlandau — CHANGES_REQUESTED** [review]
One nit, otherwise looks good

**Comment by thesamesam** [comment]
Might be worth looking at `abidiff` and the other tools from libabigail at some point.

**Review by t8m — APPROVED** [review]
Please use `git commit --fixup` instead of force pushes when fixing things from reviews.

**Review by hlandau — APPROVED** [review]

**Comment by hlandau** [comment]
> Might be worth looking at `abidiff` and the other tools from libabigail at some point.

Thanks for the suggestion! I wasn't aware of these tools. I've created an issue #22571.

**Comment by t8m** [comment]
Merged to the master branch. Thank you.

**Pull request closed without merging**